core/src/main/scala/org/bykn/bosatsu/LetFree.scala

```diff
@@ -129,42 +180,69 @@ object LetFreeConversion {
   type LetFreePM = PackageMap.Typed[(Declaration, LetFreeExpressionTag)]
   type LetFreePac = Package.Typed[(Declaration, LetFreeExpressionTag)]
 
-  type PatternEnv = Map[Int, LetFreeExpression]
+  type PatternEnv[T] = Map[Int, T]
 
   sealed trait PatternMatch[+A]
   case class Matches[A](env: A) extends PatternMatch[A]
   case object NoMatch extends PatternMatch[Nothing]
   case object NotProvable extends PatternMatch[Nothing]
 
-  val noop: (LetFreeExpression, PatternEnv) => PatternMatch[PatternEnv] = { (_, env) => Matches(env) }
-  val neverMatch: (LetFreeExpression, PatternEnv) => PatternMatch[Nothing] = { (_, _) => NoMatch }
+  def noop[T]: (T, PatternEnv[T]) => PatternMatch[PatternEnv[T]] = { (_, env) => Matches(env) }
+  def neverMatch[T]: (T, PatternEnv[T]) => PatternMatch[Nothing] = { (_, _) => NoMatch }
 
-  def structListAsList(lfe: LetFreeExpression): PatternMatch[List[LetFreeExpression]] = {
+  def structListAsList(lfe: LetFreeExpression): Option[List[LetFreeExpression]] = {
     lfe match {
-      case LetFreeExpression.Struct(0, _) => Matches(Nil)
-      case LetFreeExpression.Struct(1, List(value, tail)) => structListAsList(tail) match {
-        case Matches(lst) => Matches(value :: lst)
-        case notMatch => notMatch
-      }
-      case _ => NotProvable
+      case LetFreeExpression.Struct(0, _, _) => Some(Nil)
+      case LetFreeExpression.Struct(1, List(value, tail), _) => structListAsList(tail).map(value :: _)
+      case _ => None
     }
   }
 
   def listAsStructList(lst: List[LetFreeExpression]): LetFreeExpression =
-    lst.foldRight(LetFreeExpression.Struct(0, Nil)) { case (lfe, acc) => LetFreeExpression.Struct(1, List(lfe, acc)) }
+    lst.foldRight(LetFreeExpression.Struct(0, Nil, DataFamily.Enum)) { case (lfe, acc) => LetFreeExpression.Struct(1, List(lfe, acc), DataFamily.Enum) }
+
+  case class LitValue(toAny: Any) {
+    def equivToLit(lit: Lit) = lit match {
```

Comment:
I think `def equivToLit(lit: Lit): Boolean = toAny == lit.unboxToAny` is the same.